keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,843 @@
+
+3. Other devices consuming available slots:
+   - Network interfaces taking up slots.
+   - GPU or specialized hardware attachments that weren't present during CSI driver initialization.
+
+These scenarios can lead to the CSI driver reporting an initial capacity that becomes inaccurate over time, causing the scheduler to make decisions based on outdated information. This results in pods being scheduled to nodes without sufficient capacity, ultimately getting stuck in a `ContainerCreating` state.
+
+By making the `CSINode.Spec.Drivers[*].Allocatable.Count` field mutable and introducing a mechanism to update it dynamically, we can ensure that the scheduler always has information which more accurately represents the actual state of the world, significantly improving the reliability of stateful pod scheduling.
+
+### Goals
+
+- Make `CSINode.Spec.Drivers[*].Allocatable.Count` mutable.
+- Enable CSI drivers to define the interval at which the `Allocatable.Count` value on each node is updated through the `CSIDriver` object.
+- Automatically update `CSINode.Spec.Drivers[*].Allocatable.Count` upon detecting a failure in volume attachment due to insufficient capacity.
+
+### Non-Goals
+
+- Modifying the core scheduling logic of Kubernetes.
+- Implementing cloud provider-specific solutions within Kubernetes core.
+- Re-scheduling pods stuck in a `ContainerCreating` state.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I want the reported attachment capacity on nodes to accurately reflect the actual capacity, so that stateful pods are reliably scheduled and do not become stuck in a `ContainerCreating` state due to insufficient capacity.
+
+#### Story 2
+
+As a cluster operator, I use volumes during node setup for components like kubelet, containerd, and additional drivers. These boot volumes, which are not managed by CSI, may be detached after setup, and I need a way to reclaim these slots for other uses. The current static capacity reporting doesn't allow for this flexibility.
+
+#### Story 3
+
+As a cluster operator, I need the Kubernetes scheduler to accurately count the number of available device slots for both storage volumes and network interfaces. On certain machine types, network interfaces and volumes share device slots, and network interfaces may be dynamically
+attached after the CSI driver is registered. This results in an inaccurate `Allocatable.Count` for volumes, causing stateful
+pods to be scheduled on nodes with insufficient capacity, ultimately getting stuck in a `ContainerCreating` state.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
I would also add that it's still possible for the scheduler to race with the updates. What are the strategies a user could do to mitigate that?

Reply:
You're right, this is an important point to address, and its the reason why there are two parts to this proposal.

The periodic sync that users can opt-in to by setting `NodeAllocatableUpdatePeriodSeconds` in the `CSIDriver` object aims to generally help reduce the occurrence of this issue and also provides a mechanism by which slots could be claimed back (for example there are drivers which take into account out of band volumes and other devices during plugin registration, but once those devices are detached at a later point in time, the nodes are basically indefinitely underutilized). 

As you pointed out though, it is still possible for the scheduler to race with any updates to this value. And to be clear, that race will _always_ exist, as long as the scheduler relies on the value of `CSINode.Allocatable` to define the upper bound for attachments on nodes. But its worth highlight that even if the race is lost occasionally, the user experience is vastly improved by periodically syncing this value because today, once the value is inaccurate for any given node, the scheduler continues to make poor scheduling choices indefinitely.

> What are the strategies a user could do to mitigate that?

I believe that we can provide a much better user experience if the CO itself mitigates this race. Hence why this KEP also proposes that k8s attempt to update the value once an invalid state is detected and the race is lost (or even if there is no race at all and `NodeAllocatableUpdatePeriodSeconds` is disabled because users haven't opted in to the periodic sync).

There aren't many pragmatic solutions users could adopt to mitigate this issue on their end, and the failure mode is quite painful to recover from. What most providers have been doing is allowing users to "reserve slots" as a way to work around this (subtract x from the value that would be reported by a driver's `NodeGetInfoResponse.MaxVolumesPerNode`).